### DISCUSSION TITLE
May i know if anyone encoutner this error when fresh installnetbox 2.10.8 ?

### DISCUSSION THREAD

**Discussion by richard063** [netbox-community/netbox#13607] [Help Wanted!]
Hello all.  i have the db dump from netbox version 2.10.8 and i exported it out, so, i can setup another separate system to update it, so, it wont affect anything now in production, thats my intend.

so, now this is my plant.

1.) install netbox 2.10.8 fresh on aws EC2 instance.
2.) restore the backup *.dump database from my production netbox which is running 2.10.8\
3,) once the restore completed, update the newly setup to 2.11 then move to 3.x version.

but NOW my problem is... my 2.10.8 is not even install.

i hit this error 

ImportError: cannot import name 'soft_unicode' from 'markupsafe' (/opt/netbox-2.10.8/venv/lib/python3.8/site-packages/markupsafe/__init__.py)

as i guess is was because soft_unicide has been remove, so, may i know if anyone encouner this issue and how can i seek help to resolve this issue?

==> command i ran --> sudo /opt/netbox/upgrade.sh
output.
.
.
  File "/opt/netbox-2.10.8/venv/lib/python3.8/site-packages/jinja2/filters.py", line 13, in <module>
    from markupsafe import soft_unicode
**ImportError: cannot import name 'soft_unicode' from 'markupsafe' (/opt/netbox-2.10.8/venv/lib/python3.8/site-packages/markupsafe/__init__.py)**

i did installed the 2.0.1 version as what i found online, but it doesnt seems to help as the same error still appear. because its still showing 2.1.3!!

**pip show markupsafe**
Name: MarkupSafe
**Version: 2.1.3**
Summary: Safely add untrusted strings to HTML/XML markup.
Home-page: https://palletsprojects.com/p/markupsafe/
Author: 
Author-email: 
License: BSD-3-Clause
Location: /opt/netbox-2.10.8/venv/lib/python3.8/site-packages
Requires: 
Required-by: Jinja2

i did my homework google it and also search in this forum, but i dont seems to find anything. either no one got this issue OR i missed it. can someone please help?

**Comment by richard063** [comment]
Ah! multiple delete and install on markupsafe do the job... i have no idea why is that.

what i did was multiple removal of markupsafe and install 2.0.1 and eventually it work

**Comment by candlerb** [comment]
There is no need to install v2.10.8.  Install v2.11.12, restore the database into that, run `upgrade.sh` to migrate it.  Then you *should* be able to install v3.5.9 and run `upgrade.sh` again, job done.

However, you should read the release notes for all the major releases in between - i.e. v2.11.0, v3.0.0, v3.1.0 etc. They'll explain breaking changes, and in some cases data won't migrate if it would be lost (you'll either have to move it by hand into new models and remove from the old place, or set an environment variable to allow it to be deleted)

Unfortunately, since Netbox doesn't pin its dependencies, old versions of Netbox don't necessarily install successfully - in 2021 it may have pulled in a certain set of dependencies that worked, but installing the *same* version in 2023 may pull in a different set of dependencies, and fail.  Your problem with markupsafe is just one example of this.

To help with that problem, someone created a git repo of "pip freeze" output which should allow you to install historical versions of Netbox successfully, back to v2.11.12 anyway.

(Digs around to find it)

Here you go: https://github.com/markkuleinio/netbox-freezes

**Reply by candlerb** [reply]
Also, you will simplify your life if you use python 3.8, at least for the migration.  Netbox >= v3.3.0 doesn't support anything older, and Netbox v2.11 may have problems with anything newer.

**Reply by candlerb** [reply]
See also #11253
